### ISSUE TITLE
Quanteda installation: invalid permissions

### ISSUE THREAD

**Issue by gulmira19** [quanteda/quanteda#2342]
## Describe the bug

I have an issue with calling the library quanteda. Whenever I call this library, R issues  [*** caught segfault ***
address 0x245, cause 'invalid permissions.']

## Reproducible code:
[> install.packages("quanteda")
--- Please select a CRAN mirror for use in this session ---
trying URL 'https://cran.case.edu/bin/macosx/big-sur-arm64/contrib/4.3/quanteda_3.3.1.tgz'
Content type 'application/x-gzip' length 4232594 bytes (4.0 MB)
==================================================
downloaded 4.0 MB

The downloaded binary packages are in
	/tmp/Rtmp1YGLIE/downloaded_packages
> library(quanteda)
Package version: 3.3.1
Unicode version: 14.0
ICU version: 71.1

 *** caught segfault ***
address 0x245, cause 'invalid permissions'

Traceback:
 1: RcppParallel::defaultNumThreads()
 2: get_threads()
 3: unname(min(get_threads(), na.rm = TRUE))
 4: get_options_default()
 5: quanteda_initialize()
 6: quanteda_options(initialize = TRUE)
 7: fun(libname, pkgname)
 8: doTryCatch(return(expr), name, parentenv, handler)
 9: tryCatchOne(expr, names, parentenv, handlers[[1L]])
10: tryCatchList(expr, classes, parentenv, handlers)
11: tryCatch(fun(libname, pkgname), error = identity)
12: runHook(".onAttach", ns, dirname(nspath), nsname)
13: attachNamespace(ns, pos = pos, deps, exclude, include.only)
14: doTryCatch(return(expr), name, parentenv, handler)
15: tryCatchOne(expr, names, parentenv, handlers[[1L]])
16: tryCatchList(expr, classes, parentenv, handlers)
17: tryCatch({    attr(package, "LibPath") <- which.lib.loc    ns <- loadNamespace(package, lib.loc)    env <- attachNamespace(ns, pos = pos, deps, exclude, include.only)}, error = function(e) {    P <- if (!is.null(cc <- conditionCall(e)))         paste(" in", deparse(cc)[1L])    else ""    msg <- gettextf("package or namespace load failed for %s%s:\n %s",         sQuote(package), P, conditionMessage(e))    if (logical.return && !quietly)         message(paste("Error:", msg), domain = NA)    else stop(msg, call. = FALSE, domain = NA)})
18: library(quanteda)

Possible actions:
1: abort (with core dump, if enabled)
2: normal R exit
3: exit R without saving workspace
4: exit R saving workspace
Selection: ]

## Expected behavior

I expected that after the successful quanteda installation, the library will be called with no issue.

## System information:
#Please run `sessionInfo()` and paste the output:

[R version 4.3.2 (2023-10-31)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Sonoma 14.0

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
[1] compiler_4.3.2
> 
]
## Additional info

I tried to reinstall qunteda, updated all packages,  restarted R session, updated new R version, opened R with advanced priviliges through terminal using "sudo" command. It has been my first package I imported, so there should be no conflict between the packages. I also checked Activity Monitor on Mac to confirm that I have enough memory, but the issue is always the same - "segfault invalid permissions"

Are there any solutions?

**Comment by koheiw** [comment]
In the upcoming version, `RcppParallel::defaultNumThreads()`, which is causing the error, is not used. Can you install from Github and test?

**Comment by gulmira19** [comment]
Hello, koheiw!
Thank you for your suggestions! I have installed "RcppCore/RcppParallel" from the Github. After that, I have been able to successfully install and call quanteda. However, when I tried to perform tokenization, I faced the same seg fault:

_> library(RcppParallel)
> library(quanteda)
Package version: 3.3.1
Unicode version: 14.0
ICU version: 71.1
Parallel computing: 8 of 8 threads used.
See https://quanteda.io for tutorials and examples.
 
> texts <- c("I love programming in R.", "Text analysis is interesting.", "R is a powerful language.")
> corpus <- corpus(texts)
> tokens <- tokens(corpus)

 *** caught segfault ***
address 0x9ffffffe7, cause 'invalid permissions'

Traceback:
 1: qatd_cpp_tokens_select(x, type, ids, 2, padding, window[1], window[2],     startpos, endpos)
 2: tokens_select.tokens(x, ..., selection = "remove")
 3: tokens_select(x, ..., selection = "remove")
 4: tokens_remove(x, removals[["separators"]], valuetype = "regex",     verbose = FALSE)
 5: tokens.tokens(result, remove_punct = remove_punct, remove_symbols = remove_symbols,     remove_numbers = remove_numbers, remove_url = remove_url,     remove_separators = remove_separators, split_hyphens = FALSE,     split_tags = FALSE, include_docvars = TRUE, padding = padding,     verbose = verbose)
 6: tokens.corpus(corpus)
 7: tokens(corpus)

Possible actions:
1: abort (with core dump, if enabled)
2: normal R exit
3: exit R without saving workspace
4: exit R saving workspace
Selection:_ 

I then tried to reduce the number of threads to 1. However, it still remained equal to 8, and the issue with the seg fault persisted:

_> library(RcppParallel)
> num_threads <- RcppParallel::defaultNumThreads()
> cat("Default number of threads:", num_threads, "\n")
Default number of threads: 8 
> RcppParallel::setThreadOptions(numThreads = 1)
> actual_num_threads <- RcppParallel::defaultNumThreads()
> num_threads_after_setting = RcppParallel::defaultNumThreads()
> cat("Number of threads after setting:", num_threads_after_setting, "\n")
Number of threads after setting: 8_ 

Do you have any further ideas to resolve this issue? Thank you.

**Comment by koheiw** [comment]
Your quanteda is still v3x, please install it from Github too.
```
# remotes package required to install quanteda from Github 
remotes::install_github("quanteda/quanteda") 
```

**Comment by gulmira19** [comment]
koheiw, thank you so much for your prompt response and valuable suggestion. I tried to install quanteda from GitHub, but I received a warning message about the non-zero exit status that prevented the installation:

Downloading GitHub repo quanteda/quanteda@HEAD
── R CMD build 
  building ‘quanteda_4.0.0.tar.gz’ation ...x/l8408syn1x95x6wvrp35d9g00000gn/T/RtmpM7FzQB/remotesef067af610b/quanteda-quanteda-cb80e23/DESCRIPTION’ ...

[...]

Warning message:
In i.p(...) :
  installation of package ‘/var/folders/bx/l8408syn1x95x6wvrp35d9g00000gn/T//RtmpWKR3pI/filedbb6b127bb6/quanteda_4.0.0.tar.gz’ had non-zero exit status

**Comment by koheiw** [comment]
Do you have all those tools installed to compile the code on your machine?

https://cran.r-project.org/bin/macosx/

**Comment by gulmira19** [comment]
Hello again, koheiw! I'm sorry it took me longer to follow your suggestion and respond. I appreciate your assistance!  

I've installed R 4.3.2 binary for macOS 11 and XQuartz. When I tried to install binaries "Big Sur" for arm64-based Macs, I could not find quanteda in the list of packages "contrib" - I hope it's not a cause of a problem.

After that, I reinstalled quanteda from the GitHub again and faced the same warning message as I wrote above. 

If there specific binaries or tools from your link above that I need to install? Thanks a lot.

**Comment by koheiw** [comment]
@kbenoit do you have any idea?

**Comment by kbenoit** [comment]
1) Can you paste the first three lines of output from starting up R? e.g.

```
R version 4.3.2 (2023-10-31) -- "Eye Holes"
Copyright (C) 2023 The R Foundation for Statistical Computing
Platform: aarch64-apple-darwin20 (64-bit)

```

2) what is the output of `.libPaths()`? e.g.

```
> .libPaths()
[1] "/Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/library"
```

sometimes conflicting installations can exist if you have a local path defined as well.

As for there not being an arm64 binary, that's just not the case. There is definitely that binary built and on CRAN.

**Comment by gulmira19** [comment]
Hello, @kbenoit ! Thank you for your assistance.

1. R version 4.3.2 (2023-10-31) -- "Eye Holes"
Copyright (C) 2023 The R Foundation for Statistical Computing
Platform: aarch64-apple-darwin20 (64-bit)

2. [1] "/Library/Frameworks/R.framework/Versions/3.6/Resources/library"
Can it cause a problem if I have an older library path and a newer version of R?

**Comment by kbenoit** [comment]
Yes, I suspect that is the problem. Check your .R / .Rprofile etc files to see if that libpath is hardwired. If you remove the hard path reference, then packages will install to your versioned library automatically.

**Comment by gulmira19** [comment]
Hello, @kbenoit and @koheiw 

Thank you so much for your guidance on this issue. 

I removed the hard path reference and updated R to the latest version. I also downloaded R Studio (previously, I used R console only). Luckily, there has been no issue with quanteda on R Studio - I am very happy about it!!!

> install.packages("quanteda")
trying URL 'https://cran.rstudio.com/bin/macosx/big-sur-arm64/contrib/4.3/quanteda_3.3.1.tgz'
Content type 'application/x-gzip' length 4232594 bytes (4.0 MB)
==================================================
downloaded 4.0 MB


The downloaded binary packages are in
	/var/folders/bx/l8408syn1x95x6wvrp35d9g00000gn/T//RtmpZGYJkn/downloaded_packages
> library(quanteda)
Package version: 3.3.1
Unicode version: 14.0
ICU version: 71.1
Parallel computing: 8 of 8 threads used.
See https://quanteda.io for tutorials and examples.
> txt <- c(doc1 = "A sentence, showing how tokens() works.",
+          doc2 = "@quantedainit and #textanalysis https://example.com?p=123.",
+          doc3 = "Self-documenting code??",
+          doc4 = "£1,000,000 for 50¢ is gr8 4ever \U0001f600")
> tokens(txt)
Tokens consisting of 4 documents.
doc1 :
 [1] "A"        "sentence" ","        "showing"  "how"      "tokens"   "("       
 [8] ")"        "works"    "."       

doc2 :
[1] "@quantedainit"              "and"                       
[3] "#textanalysis"              "https://example.com?p=123."

doc3 :
[1] "Self-documenting" "code"             "?"                "?"               

doc4 :
[1] "£"         "1,000,000" "for"       "50"        "¢"         "is"       
[7] "gr8"       "4ever"     "😀"       


However, when I tried to run the same code on R Console, it showed the same seg fault I encountered before:

> library(quanteda)
Package version: 3.3.1
Unicode version: 14.0
ICU version: 71.1
Parallel computing: 8 of 8 threads used.
See https://quanteda.io for tutorials and examples.
> txt <- c(doc1 = "A sentence, showing how tokens() works.",
+          doc2 = "@quantedainit and #textanalysis https://example.com?p=123.",
+          doc3 = "Self-documenting code??",
+          doc4 = "£1,000,000 for 50¢ is gr8 4ever \U0001f600")
> tokens(txt)

 *** caught segfault ***
address 0x9ffffffe7, cause 'invalid permissions'

Traceback:
 1: qatd_cpp_tokens_select(x, type, ids, 2, padding, window[1], window[2],     startpos, endpos)
 2: tokens_select.tokens(x, ..., selection = "remove")
 3: tokens_select(x, ..., selection = "remove")
 4: tokens_remove(x, removals[["separators"]], valuetype = "regex",     verbose = FALSE)
 5: tokens.tokens(result, remove_punct = remove_punct, remove_symbols = remove_symbols,     remove_numbers = remove_numbers, remove_url = remove_url,     remove_separators = remove_separators, split_hyphens = FALSE,     split_tags = FALSE, include_docvars = TRUE, padding = padding,     verbose = verbose)
 6: tokens.corpus(corpus(x), what = what, remove_punct = remove_punct,     remove_symbols = remove_symbols, remove_numbers = remove_numbers,     remove_url = remove_url, remove_separators = remove_separators,     split_hyphens = split_hyphens, split_tags = split_tags, include_docvars = include_docvars,     padding = padding, verbose = verbose, ...)
 7: tokens.character(txt)
 8: tokens(txt)

Possible actions:
1: abort (with core dump, if enabled)
2: normal R exit
3: exit R without saving workspace
4: exit R saving workspace
Selection: 


Still, I  am glad that it works well on R studio. Any thoughts on why R Studio is doing a better job than R Console?
Thanks.

**Comment by kbenoit** [comment]
For whatever reason, R console and RStudio are reading different environment variables that determine where your packages are installed. You can compare both using `.libPaths()` within each instance of running R (console v. RStudio). Check your various .R files to make sure there is not a different one somewhere that is affecting R console's R versus Rstudio.

**Comment by gulmira19** [comment]
@kbenoit ,
Thank you for your quick response. I checked .libPaths() for both R Studio and R Console, and now I have achieved consistency between the R Studio and R Console. In both cases, the packages are installed in my versioned library. Although the issue still persists on R Console, I am happy that quanteda works seamlessly in R Studio. I could have done more if I had more technical expertise, so I am closing this thread. Anyway, your support is greatly appreciated, and I am grateful for the prompt assistance! Wishing you the best in your endeavors.
